pramen/extras/src/main/scala/za/co/absa/pramen/extras/sink/EnceladusUtils.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2022 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.pramen.extras.sink
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.functions.col
+import org.slf4j.LoggerFactory
+import za.co.absa.pramen.extras.query.QueryExecutor
+import za.co.absa.pramen.extras.utils.FsUtils
+import za.co.absa.pramen.extras.utils.PartitionUtils.unpackCustomPartitionPattern
+
+import java.time.LocalDate
+import scala.util.matching.Regex
+import scala.util.{Failure, Success, Try}
+
+class EnceladusUtils(rawPartitionPattern: String,
+                     publishPartitionPattern: String,
+                     infoDateColumn: String)
+                    (implicit spark: SparkSession,
+                     queryExecutor: QueryExecutor) {
+  private val log = LoggerFactory.getLogger(this.getClass)
+
+  val rawVersionRegEx: Regex = "^.*v(\\d+)$".r
+  val publishVersionRegEx: Regex = "^.*=(\\d+)$".r
+
+  def getNextEnceladusVersion(infoDate: LocalDate,
+                              rawBasePath: Path,
+                              publishBasePathOpt: Option[Path],
+                              hiveTableOpt: Option[String]): Try[Int] = {
+    val maxVersionInPublish = publishBasePathOpt match {
+      case Some(publishBasePath) =>
+        log.info(s"Detecting info version from the publish base path: $publishBasePath")
+        getMaxVersionInPublish(publishBasePath, infoDateColumn, infoDate)
+      case None                  =>
+        hiveTableOpt match {
+          case Some(hiveTable) =>
+            log.info(s"Detecting info version from the hive table: $hiveTable")
+            getMaxVersionFromHive(hiveTable, infoDateColumn, infoDate)
+          case None            =>
+            Failure(new IllegalArgumentException(s"No publish path or hive table specified for $rawBasePath."))
+        }
+    }
+
+    maxVersionInPublish.map {
+      case Some(versionInPublish) =>
+        versionInPublish + 1
+      case None                   =>
+        1
+    }
+  }
+
+  def getMaxVersionInRaw(rawBasePath: Path,
```

Comment:
I noticed this method is not used yet (but I could be wrong, I just couldn't find the usage). I noticed that in the previous commit it was used as a fallback if no version was found.

Reply:
Good spot! Yes, initially, I thought of using max raw version and max publish version to determine the info version to write. But later I figured out that the raw version does not really matter. Kept the code for getting the maximum raw version in case the logic changes and it is needed later.